### ISSUE TITLE
Request: Add optional timeout to blocking calls

### ISSUE THREAD

**Issue by dkozel** [rpclib/rpclib#98]
I see that making a blocking call to a function when the server is not running will block forever.

```
rpc::client rpc_client{port, addr};
rpc_client.get_connection_state(); // returns "initial"
rpc_client.call("echo", "arguments"); // hangs if the server isn't running
```

This is because the [future.wait()](https://github.com/rpclib/rpclib/blob/master/include/rpc/client.inl#L8) of a blocking call will wait forever. Would you accept a PR adding the ability to set a timeout? If so, what would your preferred response to an error be? Throwing an rpc_error with the `what` indicating timeout?

**Comment by dkozel** [comment]
I need to spend some more time looking at and using the code before I make more statements. The [wait_conn()](https://github.com/rpclib/rpclib/blob/master/lib/rpc/client.cc#L155) function is where the above code blocks.

If the server disconnects after the call to [wait_conn()](https://github.com/rpclib/rpclib/blob/master/include/rpc/client.inl#L16) returns it seems likely that the client could also get into an indefinitely blocked state, but `wait_conn` seems like the lower hanging fruit here.

**Comment by sztomi** [comment]
Yes, both are valid points. And yes, a PR is very much welcome against the dev branch for either issue :).

What would be an appropriate interface for this in your opinion? I'm thinking that the client could have state to decide the timeout. 

```cpp
rpc::client c{port, addr};
c.set_timeout(10000); // 10 seconds, maybe 30 seconds could be default
```

I don't know if it makes sense to have a separate timeout for both the call and connection. Also, per-call is hard (no way to distinguish between an argument to the RPC function and argument to the `call`). Unless the argument is added as the first which I would not want. Maybe a new `call_timeout` function?

```cpp
c.call_timout(30000, "echo", "foo");
```

or a strong timeout type maybe:

```cpp
c.call( "echo", "foo", rpc::timeout(10000));
```

(Forgive my rambling, I'm just in a hurry and don't want to forget the ideas)

**Comment by sztomi** [comment]
On the other hand, either option would create a disparity with `async_call`.

**Comment by dkozel** [comment]
I think that a global timeout is sufficient, simple, and consistent. I can imagine a few cases where it would be mildly nice to be able to set custom timeouts, but they are far from necessary and the extra complexity isn't worth it IMHO.

My expectation would be that an exception is thrown on a timeout, an rpc_error like other connection issues such as calling a non-existent function. :)

I'll work on a PR for this as it's an essential feature for my app that connection failures be detected in a timely manner and my current workaround of wrapping the `async_call`s isn't as good as having native support.

**Comment by sztomi** [comment]
@dkozel  Started working on this (hope you don't mind :)). 
Check out the unit test: https://github.com/rpclib/rpclib/blob/client_timeout/tests/rpc/client_test.cc#L84
Do you think this covers your use-case?

**Comment by dkozel** [comment]
@sztomi no problem!

Yes, I think that covers my case nicely. I'll test the dev branch.

**Comment by dkozel** [comment]
This is working well for my application. I get reliable timeouts without any source changes. The API is clean. Thanks very much.

**Comment by sztomi** [comment]
Thanks for taking the time to check. It's going to be released this weekend in 2.0.0 (it ended up an interface-breaking release so I have to bump the major version).

**Comment by qhaas** [comment]
What is the expected behavior if the server isn't up when a `call` is made after a generous `set_timeout` is set in the client, but said server is up and running well before the time specified?
